### PR TITLE
add config to restrict roles per namespace

### PR DESCRIPTION
**Pull request by rpahli**

### CODE REVIEW

File: pkg/service/restriction_service.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+func checkAllRolesMatch(binding *v1beta1.GcpRestrictionRoleBinding, roles []string) bool {
+	for _, role := range roles {
+		found := false
+		for _, check := range binding.Roles {
+			if checkIfMatch(check, role) {
+				found = true
+				break
+			}
+		}
+		if !found {
```
Comment (axdotl) [inline pkg/service/restriction_service.go:49]: it's superfluous, you can return `found` immediately (as it's initialized with `false` in line 42)

File: pkg/service/restriction_service.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+			return false
+		}
+	}
+	return true
+}
+
+func checkIfMatch(check string, toCheck string) bool {
```
Comment (axdotl) [inline pkg/service/restriction_service.go]: fussy: Name is a bit misleading, how about `matches`?

**Review by axdotl — CHANGES_REQUESTED** [review]

File: README.asciidoc [hunk 3]
```diff
@@ -7,7 +7,16 @@ this controller manges gcp service account over kubernetes resources.
...
+- creates gcp service accounts and creates secrets from the service account keyfile
+- handles the full lifecycle of a service account via CRD
+- keyfiles are only exists inside kubernetes and not saved outside
+- with version 0.2.0 you can restrict enabled roles per namespace via regular expressions (this feature is enabled by default; can be disabled with `DISABLE_RESTRICTION_CHECK`)
+
+
+
+=== Deployment
 
 first you need to create a GCP service account with at least the following permissions:
```
Comment (axdotl) [inline README.asciidoc]: Start with uppercase

File: README.asciidoc [hunk 4]
```diff
@@ -54,4 +64,42 @@ spec:
...
+apiVersion: gcp.kiwigrid.com/v1beta1
+kind: GcpServiceAccount
+metadata:
+  name: gcpserviceaccount-bucket-sample
+spec:
+  serviceAccountIdentifier: kube-bucket-example
+  serviceAccountDescription: kube-bucket-example
+  secretName: kube-bucket-example-secret
+  bindings:
+
```
Comment (axdotl) [inline README.asciidoc]: remove blank line

File: config/crds/gcp_v1beta1_gcpnamespacerestriction.yaml [hunk 5]
```diff
@@ -0,0 +1,52 @@
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  creationTimestamp: null
```
Comment (axdotl) [inline config/crds/gcp_v1beta1_gcpnamespacerestriction.yaml:4]: Can be omitted

File: config/samples/gcp_v1beta1_gcpnamespacerestriction.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+apiVersion: gcp.kiwigrid.com/v1beta1
+kind: GcpNamespaceRestriction
+metadata:
+  labels:
+    controller-tools.k8s.io: "1.0"
+  name: gcpnamespacerestriction-sample
+spec:
+  # Add fields here
+  foo: bar
```
Comment (axdotl) [inline config/samples/gcp_v1beta1_gcpnamespacerestriction.yaml]: This example will not work, as this is not a valid property

File: config/crds/gcp_v1beta1_gcpnamespacerestriction.yaml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+          type: string
+        kind:
+          type: string
+        metadata:
+          type: object
+        spec:
+          properties:
+            namespace:
+              type: string
+            regex:
```
Comment (axdotl) [inline config/crds/gcp_v1beta1_gcpnamespacerestriction.yaml:27]: I would prefer to get rid of this `regex` flag and handle `namespaces` and `resources` always as regular expressions.
@monotek What are your thoughts on this? Do you agree or do you've any concerns about it?

File: README.asciidoc [hunk 8]
```diff
@@ -54,4 +64,41 @@ spec:
...
+metadata:
+  labels:
+  name: gcpnamespacerestriction-sample
+spec:
+  namespace: test
+  regex: true
+  restrictions:
+  - resource: "^buckets/my-bucket-name$"
+    roles:
+    - "^roles/storage.objectAdmin$"
```
Comment (axdotl) [inline README.asciidoc]: You should escape the dot here `storage\.objectAdmin`

File: config/crds/gcp_v1beta1_gcpnamespacerestriction.yaml (hunk 7 above)
Comment (monotek) [reply config/crds/gcp_v1beta1_gcpnamespacerestriction.yaml:27]: No concerns if "^" and "$" is always added to the beginning and end of the value automatically if not exists.

**Comment by monotek** [comment]
I'd like to see the helm chart removed from the repo and link to the chart in kiwigrid/helm-charts instead via README.

**Comment by rpahli** [comment]
thanks for the hint but can we please stick too the current PR

**Comment by monotek** [comment]
> thanks for the hint but can we please stick too the current PR

sure: https://github.com/kiwigrid/gcp-serviceaccount-controller/pull/2

**Pull request merged**